Feature/Initial Interface

tobiasfamos opened the pull request.

tobiasfamos-jls committed:
Add Initial Template from Assignment

tobiasfamos-jls committed:
Add KV Store Interface and KV Store Stub for testing purpuses

tobiasfamos-jls committed:
Use updated KV Store Initializer

tobiasfamos-jls committed:
Use new KVStore Initialization Method and add Stub Test

tobiasfamos-jls committed:
Use new KVStore Initialization Method and add Stub Test

tobiasfamos-jls committed:
Add functions Open, Delete, Close, Create to KeyValueStore interface and stub

tobiasfamos-jls committed:
Change delete interface to have path

tobiasfamos-jls committed:
Change delete interface to have path

tobiasfamos-jls committed:
Add  Dummy Test

tobiasfamos-jls committed:
Add Todo for multiple test scenarios

tobiasfamos-jls committed:
Add config as argument for Create

tobiasfamos-jls committed:
new todo for test

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+const defPath = "."          // create in local directory
+
+type KeyValueStore interface {
+	Put(a1 uint64, a2 [10]byte)
```
Suggest to use the standard terms `key` and `value` for a... key-value store. :grin: 
```suggestion
	Put(key uint64, value [10]byte)
```

Also: How do we want to handle the case where the key already exists? Return an error? Or silently upsert the value?

joeftiger replied:
I'd say that would be an error.
As mentioned in the exercise pdf (or in the lecture, can't remember exactly) he told us that giving the interface more advanced methods like `upsert()` is fine, but not mandatory.

Therefore it would create a collision in case it'd get implemented later.

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+
+type KeyValueStore interface {
+	Put(a1 uint64, a2 [10]byte)
+	Get(a1 uint64) [10]byte
```
DITO
```suggestion
	Get(key uint64) [10]byte
```

Lavode commented on kv/testHelpers.go:
```diff
@@ -0,0 +1,24 @@
+package kv
+
+import "github.com/tobiasfamos/KVStore/stub"
```
This causes an import cycle which Go doesn't allow. `kv/testHelpers` requires `stub`, which in turn requires `kv`.

I suggest to move the `stub` file within the `kv` package. After all it's just a (stubbed) implementation of the KV store.
Once moved to within the `kv` package, all references to `stub` (imports etc) will have to be removed, I'll try to flag them all.
```suggestion
```

joeftiger replied:
I have a general question

How does the import statement work? I thought one had to give a path to where the resource is, in this case a local path.

I wonder whether the Github link would work at all because all files are in a special branch currently.

tobiasfamos replied:
The github link is not actually invoked. It is just the name of our Module as a whole string as far as I know

tobiasfamos replied:
Moved KVStub to KVStore.

Lavode commented on stub/kvStoreStub.go:
```diff
@@ -0,0 +1,30 @@
+package stub
```
That's the other end of the import cycle.
```suggestion
```

tobiasfamos replied:
Removed File. No need to implement suggestion

Lavode commented on stub/kvStoreStub.go:
```diff
@@ -0,0 +1,30 @@
+type KvStoreStub struct {
+}
+
+func (KvStoreStub) Put(a1 uint64, a2 [10]byte) {
```
```suggestion
func (KvStoreStub) Put(key uint64, value [10]byte) {
```

Lavode commented on stub/kvStoreStub.go:
```diff
@@ -0,0 +1,30 @@
+	fmt.Printf("Add at key %d value %s", a1, a2)
+}
+
+func (KvStoreStub) Get(a1 uint64) [10]byte {
```
```suggestion
func (KvStoreStub) Get(key uint64) [10]byte {
```

Lavode commented on kv/kvStore_test.go:
```diff
@@ -0,0 +1,56 @@
+import (
+	"fmt"
+	"github.com/stretchr/testify/assert"
+	stub "github.com/tobiasfamos/KVStore/stub"
```
Moving stub into `kv` package as per below.
```suggestion
```

Lavode commented on kv/kvStore_test.go:
```diff
@@ -0,0 +1,56 @@
+// Todo test for data structure full (elements > size)
+
+func TestKVStoreStub(t *testing.T) {
+	kvStub := new(stub.KvStoreStub)
```
Moving stub into `kv` package as per below.
```suggestion
	kvStub := new(KvStoreStub)
```

Lavode commented on kv/testHelpers.go:
```diff
@@ -0,0 +1,24 @@
+}
+
+func (StubTestHelper) GetEmptyInstance() KeyValueStore {
+	return new(stub.KvStoreStub)
```
Moving stub into `kv` package as per above.
```suggestion
	return new(KvStoreStub)
```

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+type KeyValueStore interface {
+	Put(a1 uint64, a2 [10]byte)
+	Get(a1 uint64) [10]byte
+	Create(config KvStoreConfig)
```
```suggestion
	Create(config KvStoreConfig) error
```
This might have to return an error if e.g. the parameters of the config are invalid.

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+	Put(a1 uint64, a2 [10]byte)
+	Get(a1 uint64) [10]byte
+	Create(config KvStoreConfig)
+	Open(path string)
```
```suggestion
	Open(path string) error
```
Might have to return an error as file IO involved.

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+	Get(a1 uint64) [10]byte
+	Create(config KvStoreConfig)
+	Open(path string)
+	Delete(path string)
```
Might have to return an error as file IO involved.
```suggestion
	Delete(path string) error
```

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+	Create(config KvStoreConfig)
+	Open(path string)
+	Delete(path string)
+	Close()
```
Might have to return an error as file IO involved.
```suggestion
	Close() error
```

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+
+func NewKvStoreInstance(size int, path string) (*KeyValueStore, error) {
+	if size > MaxMem || size == 0 {
+		return nil, errors.New("'size' is out of range")
```
```suggestion
		return nil, fmt.Errorf("'size' is out of range, must be above 0 and below %d", MaxMem)
```

joeftiger replied:
Should be below-or-equal in this case.

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+const MaxMem = 1 << (10 * 3) // 1 GB
+const defPath = "."          // create in local directory
```
Didn't realise what `def` stood for at first.
```suggestion
const defaultPath = "."          // create in local directory
```

Lavode commented on kv/kvStore.go:
```diff
@@ -0,0 +1,34 @@
+		return nil, errors.New("'size' is out of range")
+	}
+	if len(path) == 0 {
+		fmt.Println(path)
```
This is just going to print an empty line, `path` is empty in here.
```suggestion
```

Lavode commented on stub/kvStoreStub.go:
```diff
@@ -0,0 +1,30 @@
+	return [10]byte{10, 10, 1}
+}
+
+func (KvStoreStub) Open(path string) {
```
```suggestion
func (KvStoreStub) Open(path string) error {
  return nil
```

Lavode commented on stub/kvStoreStub.go:
```diff
@@ -0,0 +1,30 @@
+func (KvStoreStub) Open(path string) {
+}
+
+func (KvStoreStub) Create(config kv.KvStoreConfig) {
```
```suggestion
func (KvStoreStub) Create(config kv.KvStoreConfig) error {
  return nil
```

Lavode commented on stub/kvStoreStub.go:
```diff
@@ -0,0 +1,30 @@
+func (KvStoreStub) Create(config kv.KvStoreConfig) {
+}
+
+func (KvStoreStub) Delete(path string) {
```
```suggestion
func (KvStoreStub) Delete(path string) error {
  return nil
```

Lavode commented on stub/kvStoreStub.go:
```diff
@@ -0,0 +1,30 @@
+func (KvStoreStub) Delete(path string) {
+
+}
+func (KvStoreStub) Close() {
```
```suggestion
func (KvStoreStub) Close() error {
  return nil
```

Lavode reviewed (commented):
Major points:
- Suggest we use standard terms `key` and `value` for a KV store in method signatures
- As structured now it will cause an import cycle `kv` -> `stub` -> `kv`. As the `stub` package contains a stubbed implementation of the interface, which is by its nature tightly coupled to the interface, I suggest to not have it in a separate package. That will fix the import cycle, and also seems sane.
- We could consider committing the `go.sum`, this way we'll all have the exact same version of dependencies installed. You wouldn't normally do that for a library (which this kind of is), but given it's coursework this might simplify things. But I'm fine with either. :)

Assuming those changes are fine with you feel free to commit them, then I'll implement the misisng tests tomorrow morning. (And fix anything I might have broken by coding directly on Github, without actually compiling it :P)

joeftiger commented:
Lavode's comments seem very reasonable to me. 
I actually have no other comment for it as well, want to do some tests myself to get  more familiar with the language :-)

tobiasfamos committed:
Standard key and value naming

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos committed:
Standard key and value naming

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos committed:
Add Returning of Error

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos committed:
Add Returning of Error

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos committed:
Add Returning of Error

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos committed:
Add Returning of Error

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos committed:
Standard key and value naming

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos committed:
Update kv/kvStore.go

Co-authored-by: Michael Senn <michael@morrolan.ch>

tobiasfamos-jls committed:
Move KV Store Stub to KV Store file to prevent circular dependency and have file consistency.

tobiasfamos commented:
Sounds Awesome. You could also just split the tests if you both want to implement some. 

About `go.sum`, I don't know it but am open to it and I think some kind of dependency management is by no means a bad idea. 

I have changed the argument names and removed the cyclic dependency.
